### PR TITLE
feat: add log probabilities to OpenAI

### PR DESCRIPTION
**Pull request by antaz**
## Changes
- Implement log probabilities for the OpenAI provider

## Examples
The following example is a reproduction of the work in [https://cookbook.openai.com/examples/using_logprobs](url)

```ruby
require "llm"

def prompt(headline)
  "You will be given a headline of a news article.
  Classify the article into one of the following categories: Technology, Politics, Sports, and Art.
  Return only the name of the category, and nothing else.
  MAKE SURE your output is one of the four categories stated.
  Article headline: #{headline}"
end

headlines = [
    "Tech Giant Unveils Latest Smartphone Model with Advanced Photo-Editing Features.",
    "Local Mayor Launches Initiative to Enhance Urban Public Transport.",
    "Tennis Champion Showcases Hidden Talents in Symphony Orchestra Debut",
]

# Categorizing headlines without any confidence levels 
headlines.each do |headline|
  completion = LLM.openai(OPENAI_KEY).complete(LLM::Message.new("user", prompt(headline)))
  puts "#{headline} - #{completion.choices.first.content}"
end

# Categorizing headlines with confidence levels through log probabilities (Note the last headline)
headlines.each do |headline|
  completion = LLM.openai(OPENAI_KEY).complete(LLM::Message.new("user", prompt(headline)), logprobs: true, top_logprobs: 2)
  logprobs = completion.choices.first.logprobs["content"]
  logprobs.first["top_logprobs"].each_with_index do |p, i|
    probability = Math.exp(p['logprob']).round(2) * 100
    puts "Token #{i}: #{p['token']}"
    puts "  Logprobs: #{p['logprob']}"
    puts "  Linear probability: #{probability}%"
    puts "\n"
  end
end

```

### CODE REVIEW

File: lib/llm/message.rb [hunk 1]
```diff
@@ -2,11 +2,12 @@
 
 module LLM
   class Message
-    attr_accessor :role, :content
+    attr_accessor :role, :content, :logprobs, :top_logprobs
 
-    def initialize(role, content)
+    def initialize(role, content, logprobs: nil)
```
Comment (0x1eef) [inline lib/llm/message.rb:7]: Why don't we just pass the completion into the message ? Then we can build features on top of it, including `logprobs` and more. Plus the response would be available to the caller / conversations. Picking and choosing bits of the response doesn't seem as useful. We're essentially hiding information from the caller this way, and only providing these select bits of information that we choose. No bueno.

File: lib/llm/message.rb [hunk 2]
```diff
@@ -4,9 +4,14 @@ module LLM
   class Message
     attr_accessor :role, :content
 
-    def initialize(role, content)
+    def initialize(role, content, extra = nil)
```
Comment (0x1eef) [inline lib/llm/message.rb:7]: This should default to an empty hash (`extra = {}`)

**Review by 0x1eef — APPROVED** [review]
Looks good but we should change the default from `nil` to `{}`

File: lib/llm/message.rb (hunk 2 above)
Comment (antaz) [reply lib/llm/message.rb:7]: Thanks! I just missed that.

File: lib/llm/message.rb [hunk 3]
```diff
@@ -4,9 +4,14 @@ module LLM
...
     attr_accessor :role, :content
 
-    def initialize(role, content)
+    def initialize(role, content, extra = {})
       @role = role
       @content = content
+      @extra = extra
+    end
+
+    def logprobs
```
Comment (0x1eef) [inline lib/llm/message.rb:13]: @antaz 
Can we link to the relevant OpenAI docs that document the structure of this response ? And discuss logprobs generally ?

Comment (antaz) [reply lib/llm/message.rb:13]: @0x1eef Yes! is there a good practice for documenting something like that?

Comment (antaz) [reply lib/llm/message.rb:13]: we can use @see i guess? in Message#logprobs?

Comment (0x1eef) [reply lib/llm/message.rb:13]: Yep, with `@see`

File: lib/llm/message.rb [hunk 4]
```diff
@@ -4,9 +4,18 @@ module LLM
...
       @role = role
       @content = content
+      @extra = extra
+    end
+
+    ##
+    # @see https://platform.openai.com/docs/api-reference/chat/object#chat/object-choices logprobs docs
+    # Returns log probabilities
+    # @return [Hash]
+    def logprobs
```
Comment (0x1eef) [inline lib/llm/message.rb:17]: @antaz Can we add a `@note` ? Can we also link to the cookbook: https://cookbook.openai.com/examples/using_logprobs ?
```ruby
# @note logprobs are specific to OpenAI 
```

Comment (antaz) [reply lib/llm/message.rb:17]: Yup

**Review by 0x1eef — APPROVED** [review]
:rocket:

**Pull request merged**